src/coreclr/debug/createdump/jsonwriter.cpp

```diff
@@ -0,0 +1,159 @@
+
+JsonWriter::JsonWriter()
+{
+    m_fd = -1;
+    m_indent = JSON_INDENT_VALUE;
+    m_comma = false;
+}
+
+JsonWriter::~JsonWriter()
+{
+    if (m_fd != -1)
+    {
+        close(m_fd);
+        m_fd = -1;
+    }
+}
+
+void JsonWriter::Write(std::string& text)
+{
+    if (!DumpWriter::WriteData(m_fd, (void*)text.c_str(), text.length()))
+    {
+        throw std::exception();
+    }
+}
+
+void JsonWriter::Write(const char* buffer)
+{
+    std::string text(buffer);
+    Write(text);
+}
+
+void JsonWriter::Indent(std::string& text)
+{
+    assert(m_indent >= 0);
+    text.append(m_indent, ' ');
+}
+
+void JsonWriter::WriteSeperator(std::string& text)
+{
+    if (m_comma)
+    {
+        text.append(1, ',');
+        text.append(1, '\n');
+    }
+    Indent(text);
+}
+
+void JsonWriter::OpenBlock(const char* name, char marker)
+{
+    std::string text;
+    WriteSeperator(text);
+    if (name != nullptr)
+    {
+        text.append("\"");
+        text.append(name);
+        text.append("\" : ");
+    }
+    text.append(1, marker);
+    text.append(1, '\n');
+    m_comma = false;
+    m_indent += JSON_INDENT_VALUE;
+    Write(text);
+}
+
+void JsonWriter::CloseBlock(char marker)
+{
+    std::string text;
+    text.append(1, '\n');
+    assert(m_indent >= JSON_INDENT_VALUE);
+    m_indent -= JSON_INDENT_VALUE;
+    Indent(text);
+    text.append(1, marker);
+    m_comma = true;
+    Write(text);
+}
+
+bool JsonWriter::OpenWriter(const char* fileName)
+{
+    m_fd = open(fileName, O_WRONLY|O_CREAT|O_TRUNC, 0664);
+    if (m_fd == -1)
+    {
+        fprintf(stderr, "Could not create json file %s: %d %s\n", fileName, errno, strerror(errno));
+        return false;
+    }
+    Write("{\n");
+    return true;
+}
+
+void JsonWriter::CloseWriter()
+{
+    assert(m_indent == JSON_INDENT_VALUE);
+    Write("\n}\n");
+}
+
+void JsonWriter::WriteValue(const char* key, const char* value)
+{
+    std::string text;
+    WriteSeperator(text);
+    text.append("\"");
+    text.append(key);
+    text.append("\" : \"");
+    text.append(value);
+    text.append("\"");
+    m_comma = true;
+    Write(text);
+}
+
+void JsonWriter::WriteValueBool(const char* key, bool value)
+{
+    WriteValue(key, value ? "true" : "false");
+}
+
+void JsonWriter::WriteValue32(const char* key, uint32_t value)
+{
+    char buffer[16];
+    snprintf(buffer, sizeof(buffer), "0x%x", value);
+    WriteValue(key, buffer);
+}
+
+void JsonWriter::WriteValue64(const char* key, uint64_t value)
+{
+    char buffer[32];
+    snprintf(buffer, sizeof(buffer), "0x%" PRIx64, value);
+    WriteValue(key, buffer);
+}
+
+void JsonWriter::OpenSection(const char* sectionName)
+{
+    OpenBlock(sectionName, '{');
+}
+
+void JsonWriter::CloseSection()
+{
+    CloseBlock('}');
+}
+
+void JsonWriter::OpenArray(const char* arrayName)
+{
+    OpenBlock(arrayName, '[');
+}
+
+void JsonWriter::CloseArray()
+{
+    CloseBlock(']');
+}
+
+void JsonWriter::OpenArrayEntry()
+{
+    OpenBlock(nullptr, '{');
```

Comment:
I obviously don't know the actual JSON terminology. OpenArrayEntry/CloseArrayEntry are used to wrap each entry in an array value.  OpenArray/CloseArray are used to write an object/value that is an array.  I was trying to keep it simple and not add anymore state to the writer and the user is very explicit.  Maybe OpenArray/CloseArray should be WriteArrayValue/CloseArrayValue?

Reply:
We need OpenArrayEntry/CloseArrayEntry because all the arrays in the crash report have multiple values/objects. Maybe I should think about them like nameless objects containing other objects.

Reply:
In JSON, you have an array of objects - objects never have names. Objects have fields that are referred to as property name, value pairs, or key value pairs.